libs/ledger/src/persistent_transaction_status_cache.cpp

```diff
@@ -0,0 +1,330 @@
+
+#include "persistent_transaction_status_cache.hpp"
+
+#include "core/serializers/main_serializer.hpp"
+#include "logging/logging.hpp"
+
+namespace fetch {
+
+namespace serializers {
+
+template <typename D>
+struct MapSerializer<fetch::ledger::StakeUpdateEvent, D>
+{
+public:
+  using Type       = fetch::ledger::StakeUpdateEvent;
+  using DriverType = D;
+
+  static uint8_t const BLOCK_INDEX = 1;
+  static uint8_t const FROM        = 2;
+  static uint8_t const AMOUNT      = 3;
+
+  template <typename Constructor>
+  static void Serialize(Constructor &map_constructor, Type const &status)
+  {
+    auto map = map_constructor(3);
+    map.Append(BLOCK_INDEX, status.block_index);
+    map.Append(FROM, status.from);
+    map.Append(AMOUNT, status.amount);
+  }
+
+  template <typename MapDeserializer>
+  static void Deserialize(MapDeserializer &map, Type &status)
+  {
+    map.ExpectKeyGetValue(BLOCK_INDEX, status.block_index);
+    map.ExpectKeyGetValue(FROM, status.from);
+    map.ExpectKeyGetValue(AMOUNT, status.amount);
+  }
+};
+
+template <typename D>
+struct MapSerializer<fetch::ledger::ContractExecutionResult, D>
+{
+public:
+  using Type       = fetch::ledger::ContractExecutionResult;
+  using StatusEnum = fetch::ledger::ContractExecutionStatus;
+  using DriverType = D;
+
+  static uint8_t const STATUS        = 1;
+  static uint8_t const CHARGE_USED   = 2;
+  static uint8_t const CHARGE_RATE   = 3;
+  static uint8_t const CHARGE_LIMIT  = 4;
+  static uint8_t const FEE_CHARGED   = 5;
+  static uint8_t const EXIT_CODE     = 6;
+  static uint8_t const STAKE_UPDATES = 7;
+
+  template <typename Constructor>
+  static void Serialize(Constructor &map_constructor, Type const &status)
+  {
+    auto map = map_constructor(7);
+    map.Append(STATUS, static_cast<uint8_t>(status.status));
+    map.Append(CHARGE_USED, status.charge);
+    map.Append(CHARGE_RATE, status.charge_rate);
+    map.Append(CHARGE_LIMIT, status.charge_limit);
+    map.Append(FEE_CHARGED, status.fee);
+    map.Append(EXIT_CODE, status.return_value);
+    map.Append(STAKE_UPDATES, status.stake_updates);
+  }
+
+  template <typename MapDeserializer>
+  static void Deserialize(MapDeserializer &map, Type &status)
+  {
+    DeserializeStatus(map, status.status);
+    map.ExpectKeyGetValue(CHARGE_USED, status.charge);
+    map.ExpectKeyGetValue(CHARGE_RATE, status.charge_rate);
+    map.ExpectKeyGetValue(CHARGE_LIMIT, status.charge_limit);
+    map.ExpectKeyGetValue(FEE_CHARGED, status.fee);
+    map.ExpectKeyGetValue(EXIT_CODE, status.return_value);
+    map.ExpectKeyGetValue(STAKE_UPDATES, status.stake_updates);
+  }
+
+  template <typename MapDeserializer>
+  static void DeserializeStatus(MapDeserializer &map, StatusEnum &status)
+  {
+    // read the raw value
+    uint8_t raw_enum_value{0};
+    map.ExpectKeyGetValue(STATUS, raw_enum_value);
+
+    // convert the raw value
+    auto converted_value = static_cast<StatusEnum>(raw_enum_value);
```

Comment:
I don't think this trick works as well as one would reasonably expect. Can you double-check this cast does not blow up if `raw_enum_value` has a value not covered by `StatusEnum`? I recall trying to do something similar back on Windows, resulting in runtime errors. To my mind, the right way would be to exclude invalid values with `if (raw_enum_value < StatusEnum::INEXPLICABLE_FAILURE) { ... }` or similar.

Reply:
```
$ cat omg.cpp 
#include <iostream>

enum class Omg {
	A,
	B
};

int main() {
	Omg x = static_cast<Omg>(3);
	std::cout << static_cast<int>(x) << '\n';
}
$ g++ -std=c++14 -Wpedantic omg.cpp 
$ ./a.out 
3
```
I don't see anything on CPPRef about casting values of underlying type to enum, even if they don't equal to any of the named constants.

Reply:
I had a look and the only restriction appears to be this:

> Values of integer, floating-point, and enumeration types can be converted by static_cast or explicit cast, to any enumeration type. If the underlying type is not fixed and the source value is out of range, the result is unspecified (until C++17)undefined (since C++17). 

https://en.cppreference.com/w/cpp/language/enum

In this case we should be ok, because `raw_enum_value` is a `uint8_t`, and `StatusEnum` is an enum class of unspecified underlying type, which in practice means its an `int`. I withdraw my question above.

Just to check I'm not mad, I had a look at MSDN. It appears that Visual Studio's enum implementation is indeed non-conformant:

> The static_cast operator can explicitly convert an integral value to an enumeration type. If the value of the integral type does not fall within the range of enumeration values, the resulting enumeration value is undefined.

https://docs.microsoft.com/en-us/cpp/cpp/static-cast-operator?view=vs-2019